Recycle serialization buffers on transmission (#342)
osrf/rmw_zenoh: docs/design.md

## Security

TBD

## Environment variables

### `RMW_ZENOH_BUFFER_POOL_MAX_SIZE_BYTES`

> The RMW recycles serialization buffers on transmission using a buffer pool with bounded memory
> usage. These buffers are returned to the pool — without being deallocated — once they cross the
> network boundary in host-to-host communication, or after transmission in inter-process
> communication, or upon being consumed by subscriptions in intra-process communication, etc. 

Reviewer: ```suggestion
The RMW recycles serialization buffers on transmission using a buffer pool with bounded memory
usage.
These buffers are returned to the pool - without being deallocated - once they cross the
network boundary in host-to-host communication, or after transmission in inter-process
communication, or upon being consumed by subscriptions in intra-process communication, etc. 
```

communication, or upon being consumed by subscriptions in intra-process communication, etc. 

When the total size of the allocated buffers within the pool exceeds

> `RMW_ZENOH_BUFFER_POOL_MAX_SIZE_BYTES`, serialization buffers are allocated using the system

Reviewer 2: What happens if we set this to `0`?

Author: The the max buffer pool size is 0 and we never allocate through it.

> allocator and moved to Zenoh; no recyling is performed in this case to prevent the buffer pool from

Reviewer: ```suggestion
allocator and moved to Zenoh; no recycling is performed in this case to prevent the buffer pool from
```

growing uncontrollably. 

> The default value of `RMW_ZENOH_BUFFER_POOL_MAX_SIZE_BYTES` is roughly proportionate to the cache
> size of a typical modern CPU.

Reviewer: ```suggestion
The default value of `RMW_ZENOH_BUFFER_POOL_MAX_SIZE_BYTES` is 16MB; this value was chosen since it is roughly the size of the cache in a modern CPU.
```

## Security

TBD

## Environment variables

> ### `RMW_ZENOH_BUFFER_POOL_MAX_SIZE_BYTES`

Reviewer 2: Add this under the Configuration section of the README instead since that section includes other environment variables one can configure? Then we can move all the configuration contents in the `README.md` to `docs/design.md` as part of https://github.com/ros2/rmw_zenoh/pull/395/files

Reviewer 3: Addressed in cefeedeacd53cc3d310bcd08605cd9fd4e3ca797.
